lib/kernel/extra.ex

```diff
@@ -67,4 +68,55 @@ defmodule Kernel.Extra do
       end
     end
   end
+
+
+  @doc """
+  Define a type union, module attribute, public function, and guard macro for a named list of
+  supported values all in one go.
+
+  Concretely, the following:
+
+      defmodule Quark do
+        import Kernel.Extra
+
+        defunion :flavor, [:up, :down, :strange, :charm, :top, :bottom]
+      end
+
+  Transforms into essentially the following:
+
+      defmodule Quark do
+        import Kernel.Extra
+
+        @flavors [:up, :down, :strange, :charm, :top, :bottom]
+
+        @type flavor :: :up | :down | :strange | :charm | :top | :bottom
+
+        def __union_flavors__, do: @flavors
+
+        defmacro is_flavor(x) do
+          quote do: unquote(x) in unquote(@flavors)
+        end
+      end
+
+  """
+  @spec defunion(name :: atom, [atom]) :: Macro.t
+  defmacro defunion(name, values) do
+    type_rhs =
+      Enum.reduce(values, fn (x, acc) -> quote do: unquote(x) | unquote(acc) end)
+
+    # XXX use real pluralization, if it's ever necessary
+    plural =
+      :"#{name}s"
+
+    quote do
+      @type unquote(Macro.var(name, nil)) :: unquote(type_rhs)
+
+      Module.put_attribute __MODULE__, unquote(plural), unquote(values)
+
+      def unquote(:"__union_#{plural}__")(), do: unquote(values)
+
+      defmacro unquote(:"is_#{name}")(x), do: {:in, [], [x, unquote(values)]}
```

Comment:
I was torn on this being public or private, but I think I like public... allowing it to be imported elsewhere is nice. That being said, it might be worth auto-generating the documentation here as well (in the quote)

Reply:
I like it public too